File: src/sim/events.ts
import type { BattleResult, BossActionType } from "./types";

/**
 * Événements émis par la simulation (patron Observer).
 * La simulation ne connaît pas ses observateurs : l'UI (sons, chiffres
 * flottants), les tests de non-régression et un futur replay serveur s'y
 * abonnent sans que `Battle` ait à changer.
 */
export type BattleEvent =
  | { type: "skillUsed"; timeMs: number; casterId: string; skillId: string; targetIds: string[] }
  | { type: "healed"; timeMs: number; unitId: string; amount: number }
  | { type: "shielded"; timeMs: number; unitId: string; amount: number }
  | { type: "unitDamaged"; timeMs: number; unitId: string; amount: number; absorbed: number }
  | { type: "bossDamaged"; timeMs: number; sourceId: string; amount: number }
  | { type: "unitDied"; timeMs: number; unitId: string }
  | { type: "bossAction"; timeMs: number; action: BossActionType; hitsAll: boolean }
  | { type: "battleEnded"; timeMs: number; result: BattleResult };

export type BattleListener = (event: BattleEvent) => void;

/** Représentation texte stable d'un événement (utilisée par les tests de non-régression). */
export function formatEvent(e: BattleEvent): string {
  const head = `${e.timeMs}ms ${e.type}`;
  switch (e.type) {
    case "skillUsed":
      return `${head} ${e.casterId} ${e.skillId} -> [${e.targetIds.join(",")}]`;
    case "healed":
    case "shielded":
      return `${head} ${e.unitId} ${e.amount}`;
    case "unitDamaged":
      return `${head} ${e.unitId} ${e.amount} (absorbé ${e.absorbed})`;
    case "bossDamaged":
      return `${head} par ${e.sourceId} ${e.amount}`;
    case "unitDied":
      return `${head} ${e.unitId}`;
    case "bossAction":
      return `${head} ${e.action}${e.hitsAll ? " zone" : ""}`;
    case "battleEnded":
      return `${head} ${e.result}`;
  }
}

File: core/Healer.Combat/Events.cs
using System;
using System.Collections.Generic;
using System.Globalization;

namespace Healer.Combat
{
    /// <summary>
    /// Événements émis par la simulation (patron Observer). La simulation ne connaît pas ses
    /// observateurs : l'UI, les sons, les tests golden et un futur replay serveur s'y abonnent.
    /// Le format texte (<see cref="Format"/>) est STABLE : il fonde les références golden.
    /// </summary>
    public sealed class BattleEvent
    {
        public string Type { get; set; } = "";
        public double TimeMs { get; set; }

        public string UnitId { get; set; } = "";
        public string CasterId { get; set; } = "";
        public string SourceId { get; set; } = "";
        public string SkillId { get; set; } = "";
        public string EffectId { get; set; } = "";
        public string Action { get; set; } = "";
        public string Reason { get; set; } = "";
        public string Result { get; set; } = "";
        public string Name { get; set; } = "";
        public IReadOnlyList<string> TargetIds { get; set; } = Array.Empty<string>();
        public double Amount { get; set; }
        public double Absorbed { get; set; }
        public int Phase { get; set; }
        public bool HitsAll { get; set; }

        /// <summary>Représentation texte identique à la version TypeScript (formatEvent).</summary>
        public string Format()
        {
            string head = $"{Num(TimeMs)}ms {Type}";
            switch (Type)
            {
                case "skillUsed":
                    return $"{head} {CasterId} {SkillId} -> [{string.Join(",", TargetIds)}]";
                case "healed":
                case "shielded":
                    return $"{head} {UnitId} {Num(Amount)}";
                case "unitDamaged":
                    return $"{head} {UnitId} {Num(Amount)} (absorbé {Num(Absorbed)})";
                case "bossDamaged":
                    return $"{head} par {SourceId} {Num(Amount)}";
                case "unitDied":
                    return $"{head} {UnitId}";
                case "bossAction":
                    return $"{head} {Action}{(HitsAll ? " zone" : "")}";
                case "effectApplied":
                    return $"{head} {UnitId} {EffectId}";
                case "effectTick":
                    return $"{head} {UnitId} {EffectId} {Num(Amount)} (absorbé {Num(Absorbed)})";
                case "effectEnded":
                    return $"{head} {UnitId} {EffectId} {Reason}";
                case "bossEnraged":
                    return $"{head} {Phase} +{Num(Amount)}%";
                case "bossPhaseChanged":
                    return $"{head} {Phase} {Name}";
                case "battleEnded":
                    return $"{head} {Result}";
                default:
                    throw new InvalidOperationException($"Type d'événement inconnu : {Type}");
            }
        }

        /// <summary>Nombre affiché comme en JavaScript : entier sans décimales, sinon valeur exacte.</summary>
        public static string Num(double value)
        {
            if (value == Math.Floor(value) && Math.Abs(value) < 1e15)
                return ((long)value).ToString(CultureInfo.InvariantCulture);
            return value.ToString("R", CultureInfo.InvariantCulture);
        }
    }
}
